packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,499 @@
+		}
+	}
+}
+
+function getSchemaIdentifier(content: TreeEditValue): string | undefined {
+	switch (typeof content) {
+		case "boolean": {
+			return booleanSchema.identifier;
+		}
+		case "number": {
+			return numberSchema.identifier;
+		}
+		case "string": {
+			return stringSchema.identifier;
+		}
+		case "object": {
+			if (content === null) {
+				return nullSchema.identifier;
+			}
+			if (Array.isArray(content)) {
+				throw new UsageError("Arrays are not currently supported in this context");
+			}
+			if (isFluidHandle(content)) {
+				return handleSchema.identifier;
+			}
+			return content[typeField];
+		}
+		default: {
+			throw new UsageError("Unsupported content type");
+		}
+	}
+}
+
+function contentWithIds(content: TreeNode, idGenerator: IdGenerator): TreeEditObject {
+	return JSON.parse(toDecoratedJson(idGenerator, content)) as TreeEditObject;
+}
+
+/**
+ * Manages applying the various types of {@link TreeEdit}'s to a a given {@link TreeNode}.
+ */
+export function applyAgentEdit(
+	treeNode: TreeNode,
+	treeEdit: TreeEdit,
+	idGenerator: IdGenerator,
+	definitionMap: ReadonlyMap<string, SimpleNodeSchema>,
+	validator?: (edit: TreeNode) => void,
+): TreeEdit {
+	objectIdsExist(treeEdit, idGenerator);
+	switch (treeEdit.type) {
+		case "insert": {
+			const { array, index } = getPlaceInfo(treeEdit.destination, idGenerator);
+
+			const parentNodeSchema = Tree.schema(array);
+			populateDefaults(treeEdit.content, definitionMap);
+
+			const schemaIdentifier = getSchemaIdentifier(treeEdit.content);
+
+			// We assume that the parentNode for inserts edits are guaranteed to be an arrayNode.
+			const allowedTypes = [
+				...normalizeAllowedTypes(parentNodeSchema.info as ImplicitAllowedTypes),
+			];
+
+			for (const allowedType of allowedTypes.values()) {
+				if (allowedType.identifier === schemaIdentifier && typeof allowedType === "function") {
+					const simpleNodeSchema = allowedType as unknown as new (dummy: unknown) => TreeNode;
+					const insertNode = new simpleNodeSchema(treeEdit.content);
+					validator?.(insertNode);
+					array.insertAt(index, insertNode);
+					return {
+						...treeEdit,
+						content: contentWithIds(insertNode, idGenerator),
+					};
+				}
+			}
+			fail("inserted node must be of an allowed type");
+		}
+		case "remove": {
+			const source = treeEdit.source;
+			if (isObjectTarget(source)) {
+				const node = getNodeFromTarget(source, idGenerator);
+				const parentNode = Tree.parent(node);
+				// Case for deleting rootNode
+				if (parentNode === undefined) {
+					const treeSchema = Tree.schema(treeNode);
+					if (treeSchema instanceof FieldSchema && treeSchema.kind === FieldKind.Optional) {
+						// eslint-disable-next-line no-param-reassign, @typescript-eslint/no-explicit-any
+						(treeNode as any) = undefined;
+					} else {
+						throw new UsageError(
+							"The root is required, and cannot be removed. Please use modify edit instead.",
+						);
+					}
```

Comment:
Based on earlier comments, is setting this treeNode to undefined not doing anything here?

Reply:
All that does is set the local `treeNode` variable to `undefined`. It doesn't affect the underlying object.

Reply:
What would be the proper way to set the root node to undefined?

Reply:
I think you need the full `TreeView` and do `view.root = undefined`. I don't recall if we can create a tree where `undefined` is a valid root node, though?

Reply:
Or if the "root" is just the root of a _subtree_ but can really be any node, then get the parent, and do `parent.<property> = undefined`. Which won't work for the actual root of the tree because it has no parent, so there you'd need the view.

Reply:
FWIW, I don't think we need this if block at all. Since we no longer support the `replace root` operation, there should be no way to remove the root node of the tree. Instead, I believe we can just assert that the parent is defined in this "remove" case.

Reply:
@alexvy86 Just FYI you can have the root field of your tree be optional, in which case undefined is totally fair game. We often do something like `new TreeConfiguration({ schema: MyRootNode })` - a required field, but it's completely possible to do `new TreeConfiguration({ schema: sf.optional(MyRootNode) })`

Reply:
BTW, based on the discussion, this has been updated to remove the ability to set the root to undefined via a `remove` TreeEdit as of commit bcf7e43a9a34491c64543194d98321fb9b54ef8d

Reply:
@seanimam - if the LLM tries to remove the root (by targeting the root node with a remove edit), the error we give is `"The root is required, and cannot be removed. Please use modify edit instead."`. But there is no way that a modify edit can change the root either, because it must specify a parent node (which the root does not have).